pkg/generate/dockercompose/generate.go

```diff
@@ -0,0 +1,537 @@
+	builds := make(map[string]*app.Pipeline)
+
+	// identify colocated components due to shared volumes
+	joins := make(map[string]sets.String)
+	volumesFrom := make(map[string][]string)
+	for _, k := range serviceOrder.List() {
+		if joins[k] == nil {
+			joins[k] = sets.NewString(k)
+		}
+		v := p.Configs[k]
+		for _, from := range v.VolumesFrom {
+			switch parts := strings.Split(from, ":"); len(parts) {
+			case 1:
+				joins[k].Insert(parts[0])
+				volumesFrom[k] = append(volumesFrom[k], parts[0])
+			case 2:
+				target := parts[1]
+				if parts[1] == "ro" || parts[1] == "rw" {
+					target = parts[0]
+				}
+				joins[k].Insert(target)
+				volumesFrom[k] = append(volumesFrom[k], target)
+			case 3:
+				joins[k].Insert(parts[1])
+				volumesFrom[k] = append(volumesFrom[k], parts[1])
+			}
+		}
+	}
+	joinOrder := sets.NewString()
+	for k := range joins {
+		joinOrder.Insert(k)
+	}
+	var colocated []sets.String
+	for _, k := range joinOrder.List() {
+		set := joins[k]
+		matched := -1
+		for i, existing := range colocated {
+			if set.Intersection(existing).Len() == 0 {
+				continue
+			}
+			if matched != -1 {
+				return nil, fmt.Errorf("%q belongs with %v, but %v also contains some overlapping elements", k, set, colocated[matched])
+			}
+			existing.Insert(set.List()...)
+			matched = i
+			continue
+		}
+		if matched == -1 {
+			colocated = append(colocated, set)
+		}
+	}
+
+	// identify service aliases
+	aliases := make(map[string]sets.String)
+	for _, v := range p.Configs {
+		for _, s := range v.Links.Slice() {
+			parts := strings.SplitN(s, ":", 2)
+			if len(parts) != 2 || parts[0] == parts[1] {
+				continue
+			}
+			set := aliases[parts[0]]
+			if set == nil {
+				set = sets.NewString()
+				aliases[parts[0]] = set
+			}
+			set.Insert(parts[1])
+		}
+	}
+
+	// find and define build pipelines
+	for _, k := range serviceOrder.List() {
+		v := p.Configs[k]
+		if len(v.Build) == 0 {
+			continue
+		}
+		if _, ok := builds[v.Build]; ok {
+			continue
+		}
+		var base, relative string
+		for _, s := range bases {
+			if !strings.HasPrefix(v.Build, s) {
+				continue
+			}
+			base = s
+			path, err := filepath.Rel(base, v.Build)
+			if err != nil {
+				return nil, fmt.Errorf("path is not relative to base: %v", err)
+			}
+			relative = path
+			break
+		}
+		if len(base) == 0 {
+			return nil, fmt.Errorf("build path outside of the compose file: %s", v.Build)
+		}
+
+		// if this is a Git repository, make the path relative
+		if root, err := git.NewRepository().GetRootDir(base); err == nil {
+			if relative, err = filepath.Rel(root, filepath.Join(base, relative)); err != nil {
+				return nil, fmt.Errorf("unable to find relative path for Git repository: %v", err)
+			}
+			base = root
+		}
+		buildPath := filepath.Join(base, relative)
+
+		// TODO: what if there is no origin for this repo?
+
+		glog.V(4).Infof("compose service: %#v", v)
+		repo, err := app.NewSourceRepositoryWithDockerfile(buildPath, "")
```

Comment:
Eventually

On Fri, Apr 29, 2016 at 1:36 PM, Cesar Wong <notifications@github.com>
wrote:

> In pkg/generate/dockercompose/generate.go
> <https://github.com/openshift/origin/pull/7690#discussion_r61616616>:
>
> > +			return nil, fmt.Errorf("build path outside of the compose file: %s", v.Build)
> > +		}
> > +
> > +		// if this is a Git repository, make the path relative
> > +		if root, err := git.NewRepository().GetRootDir(base); err == nil {
> > +			if relative, err = filepath.Rel(root, filepath.Join(base, relative)); err != nil {
> > +				return nil, fmt.Errorf("unable to find relative path for Git repository: %v", err)
> > +			}
> > +			base = root
> > +		}
> > +		buildPath := filepath.Join(base, relative)
> > +
> > +		// TODO: what if there is no origin for this repo?
> > +
> > +		glog.V(4).Infof("compose service: %#v", v)
> > +		repo, err := app.NewSourceRepositoryWithDockerfile(buildPath, "")
>
> In version 2 of compose you can specify an alternate Dockerfile and
> args/env ... do we plan to support those?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7690/files/b85cc800bdcb1e7db8f8c37dd3e85c21b5023f0a#r61616616>
>